Add a section to help users understand key SDK and core concepts with samples. (#2394)
Azure/azure-sdk-for-cpp: README.md

> cmake --build [build directory]
```

#### Using the SDK within your Application

The **entry point** for most scenarios when using the SDK will be a top-level client type corresponding to the Azure service you want to talk to. For example, sending requests to blob storage can be done via the `Azure::Storage::Blobs::BlobClient` API. All APIs on the client type send HTTP requests to the cloud service and return back an HTTP `Response<T>`.

> All the Azure C++ SDK headers needed to be included are located within the `<azure>` folder, with sub-folders corresponding to each service. Similarly, all types and APIs can be found within the `Azure::` namespace. For example, to use functionality from `Azure::Core`, include the following header at the beginning of your application `#include <azure/core.hpp>`.

Reviewer: I don'tthink users will need to include core that often in real world. Should we use <azure/storage/blobs.hpp> as example instead?

Author: I don't think service client library headers include all of core.hpp, I think adding core.hpp is likely going to be useful for end users for using things that might be used by one service but not another.

Here's an example application to help you get started:

```C++
#include <iostream>

// Include the necessary SDK headers

> #include <azure/core.hpp>

Reviewer: This is very likely not needed. Blobs should include everything they need from core.

#include <azure/storage/blobs.hpp>

// Use the appropriate namespace using directives
using namespace Azure::Storage::Blobs;

> // Get the required connection string/key from an environment variable or Azure KeyVault.

Reviewer: I think the branding is "Azure Key Vault"

Reviewer 2: correct, 
```suggestion
// Get the required connection string/key from an environment variable or Azure Key Vault.

```

Reviewer 2: Actually, why are we mentioning Key Vault at all?  The sample is just for Storage isn't it?

I would remove the entire sample code from here and leave it only inside each service Readme to sample the code.  Maybe just add links to where the samples for the current services are.

Author: That note/comment is just there on where/how to read the connection string, to indicate best practices on where to get that from, the sample doesn't have anything to do with key vault.

> std::string GetConnectionString();

Reviewer 3: Is this a TODO or something?
We actually discourage the use of connection strings; can we change this?

Author: @JeffreyRichter how do you recommend we construct a blob client, if not using a connection string, as a copy-pasteable sample for customers?

Reviewer 2: > We actually discourage the use of connection strings

Why?  For Storage, that's super easy to get the connection string from the Azure portal and use it.  Is it even easier to set the shareKeyCredential?

Reviewer 3: Connection strings usually combine multiple things together that should not be combined like endpoint, username & password. We want to encourage people to roll over passwords periodically while their service keeps running but if they are in a connection string, then they need a whole new connection string and there is no way to atomically update an endpoint, username, and password.

In addition, there are many reasons why distributed systems fail including bad endpoints and bad credentials. So, customers can better diagnose issues if they are responsible for passing these things explicitly – it is the cost of entry for building a distributed system.

Connection strings also have the potential for lots of problems like are they case sensitive? What does it mean if multiple credentials appear in the string? Will customers embed connection strings in their source code and check that in to github? (connections strings make this bad practice too easy).

Azure SDKs policy is to NOT support connection strings unless the portal/tooling does offer them. In this case, we support them for "get up and running fast" scenario but we still discourage their use for real production.

It's even in our C++ guidelines: https://azure.github.io/azure-sdk/cpp_introduction.html#cpp-authentication

All the Azure C++ SDK headers needed to be included are located within the `<azure>` folder, with sub-folders corresponding to each service. Similarly, all types and APIs can be found within the `Azure::` namespace. For example, to use functionality from `Azure::Core`, include the following header at the beginning of your application `#include <azure/core.hpp>`.

Here's an example application to help you get started:

```C++
#include <iostream>

// Include the necessary SDK headers
#include <azure/core.hpp>
#include <azure/storage/blobs.hpp>

// Use the appropriate namespace using directives
using namespace Azure::Storage;
using namespace Azure::Storage::Blobs;

// Since they are secrets, get the required account name and key string from an environment variable
// or Azure Key Vault.
// Leaving this up to the user to implement.
std::string GetAccountName();
std::string GetAccountKey();

int main()
{
  std::string endpointUrl = "<update to use your storage account url here>";
  std::string accountName = GetAccountName();

>   std::string accountKey = GetAccountKey();

Reviewer: Just a proposal: implement `GetStorageAccountUrl()`, `GetAccountName()`, `GetAccountKey()` as one-liners returning `{ return std::getenv("AZURE_STORAGE_ACCOUNT_URL"); }`, and so on.

Author: I sort of agree, but the sample is already a bit verbose, this is mainly getting started, and the main intention here is to highlight the basic concepts.

I would probably also expect the user to add error handling when the environment variable doesn't exist. The user has to modify the code to choose their own environment variable key name anyway.

Author: Alright, took your suggestion Anton, thanks!

> cmake --build [build directory]
```

#### Using the SDK within your Application

> The **entry point** for most scenarios when using the SDK will be a top-level client type corresponding to the Azure service you want to talk to. For example, sending requests to blob storage can be done via the `Azure::Storage::Blobs::BlobClient` API. All APIs on the client type send HTTP requests to the cloud service and return back an HTTP `Response<T>`.

Reviewer 4: Nit, drop this
```suggestion
The **entry point** for most scenarios when using the SDK will be a top-level client type corresponding to the Azure service. For example, sending requests to blob storage can be done via the `Azure::Storage::Blobs::BlobClient` API. All APIs on the client type send HTTP requests to the cloud service and return back an HTTP `Response<T>`.

```

> All the Azure C++ SDK headers needed to be included are located within the `<azure>` folder, with sub-folders corresponding to each service. Similarly, all types and APIs can be found within the `Azure::` namespace. For example, to use functionality from `Azure::Core`, include the following header at the beginning of your application `#include <azure/core.hpp>`.

Reviewer 4: Nit, 
```suggestion
Azure C++ SDK headers are located within the `<azure>` folder, with sub-folders corresponding to each service. Similarly, all types and APIs can be found within the `Azure::` namespace. For example, to use functionality from `Azure::Core`, include the following header at the beginning of your application `#include <azure/core.hpp>`.

```

Here's an example application to help you get started:

```C++
#include <iostream>

// Include the necessary SDK headers
#include <azure/core.hpp>
#include <azure/storage/blobs.hpp>

> // Use the appropriate namespace using directives

Reviewer 4: Nit
```suggestion
// Add appropriate using namespace directives

```

using namespace Azure::Storage;
using namespace Azure::Storage::Blobs;

> // Since they are secrets, get the required account name and key string from an environment variable

Reviewer 4: Check the formatting before accepting this

```suggestion
// Secrets should be stored & retrieved from secure locations such as Azure::KeyVault.  For convenience and brevity of samples, the secrets are retrieved from environment variables.
//  Get the required account name and key string from an environment variable

```

// variable or Azure Key Vault. Leaving this up to the user to implement.

> std::string GetEndPointUrl();

Reviewer: ```suggestion
std::string GetEndpointUrl();

```

std::string GetAccountName();
std::string GetAccountKey();

int main()
{

>   std::string endpointUrl = "<update to use your storage account url here>";

Reviewer 2: ```suggestion
  std::string endpointUrl = GetEndPointUrl();

```

Reviewer: Do not break the word, use `Endpoint` instead of `EndPoint`

>   std::string endpointUrl = GetEndPointUrl();

Reviewer: ```suggestion
  std::string endpointUrl = GetEndPointUrl();
```suggestion
  std::string endpointUrl = GetEndpointUrl();

```
```

  std::string accountName = GetAccountName();
  std::string accountKey = GetAccountKey();

  try
  {

>     // There are many types of client types, use the one that's appropriate

Reviewer 4: I don't find this comment very helpful.  I suggest removing.

Reviewer: BTW, should we keep it, I think it is better to rephrase as "there are many types of clients"

    // for your scenario, such as BlockBlobClient.
    auto sharedKeyCredential = std::make_shared<StorageSharedKeyCredential>(
        StorageSharedKeyCredential(accountName, accountKey));

    auto blockBlobClient = BlockBlobClient(endpointUrl, sharedKeyCredential);

>     // Assuming there is some Azure::Core::IO::BodyStream containing the data to upload.

Reviewer 4: No assumption here, it has some random data that we are uploading.  I'd state "Create some data to upload into the blob"

    // For example either a MemoryBodyStream or FileBodyStream.
    std::vector<uint8_t> data = {1, 2, 3, 4};
    Azure::Core::IO::MemoryBodyStream stream(data);

    Azure::Response<Models::UploadBlockBlobResult> response = blockBlobClient.Upload(stream);

    Models::UploadBlockBlobResult model = response.Value;
    std::cout << "Last modified date of uploaded blob: " << model.LastModified.ToString()
              << std::endl;
  }
  catch (const Azure::Core::RequestFailedException& e)
  {
    std::cout << "Status Code: " << static_cast<int>(e.StatusCode)
              << ", Reason Phrase: " << e.ReasonPhrase << std::endl;
    std::cout << e.what() << std::endl;
    return 1;
  }
  return 0;
}
```

#### Key Core concepts

> Understanding the key concepts from the `Azure Core` library, which is leveraged by all client libraries will also be helpful in getting started, regardless of which Azure service you want to use.

Reviewer 4: ```suggestion
Understanding the key concepts from the `Azure Core` library, which is leveraged by all client libraries is helpful in getting started, regardless of which Azure service you want to use.

```

The main shared concepts of `Azure Core` include:

- Accessing HTTP response details for the returned model of any SDK client operation, via `Response<T>`.
- Exceptions for reporting errors from service requests in a consistent fashion via the base exception type `RequestFailedException`.
- Abstractions for Azure SDK credentials (`TokenCredential`).
- Handling streaming data and input/output (I/O) via `BodyStream` along with its derived types.
- Polling long-running operations (LROs), via `Operation<T>`.

> - Operations that return collections do so via a `PagedResponse<T>`.

Reviewer 4: ```suggestion
- Collections are returned via `PagedResponse<T>`.

```

```C++
  // Azure service operations return a Response<T> templated type.
  Azure::Response<Models::BlobProperties> propertiesResponse = blockBlobClient.GetProperties();

  // You can get the T, from the returned Response<T>,
  // which is typically named with a Result suffix in the type name.
  Models::BlobProperties propertiesModel = propertiesResponse.Value;

  // Now you can look at API specific members on the result object that is returned.
  std::cout << "The size of the blob is: " << propertiesModel.BlobSize << std::endl;
```

#### Long Running Operations

Some operations take a long time to complete and require polling for their status. Methods starting long-running operations return `Operation<T>` types.

You can intermittently poll whether the operation has finished by using the `Poll()` method inside a loop on the returned `Operation<T>` and track progress of the operation using `Value()`, while the operation is not done (using `IsDone()`). Your per-polling custom logic can go in that loop, such as logging progress.Alternatively, if you just want to wait until the operation completes, you can use `PollUntilDone()`.

```C++

>   std::string sourceUri = "<a uri to the source blob to copy>";

Reviewer: nit: mismatch between url (above) and url (here) usage. Also, maybe capitalize?

Author: I know. The API is called StartCopyFromUri, so that's where the inconsistency issue comes from :(

Reviewer: I see, the API is called ...FormUri, it is better to align with the API then.

  try
  {
    // There are many types of client types, use the one that's appropriate
    // for your scenario, such as BlobContainerClient.
    BlobClient client
        = BlobClient::CreateFromConnectionString(GetConnectionString(), containerName, blobName);

>     BlockBlobClient block = client.AsBlockBlobClient();

Reviewer 3: If you get rid of connection strings (which you should), then you should just new up a BlockBlobClient and not have to call AsBlockBlobClient.

Reviewer 3: I also suggest renaming the variable to blockBlobClient; not just block.

Reviewer 3: You could use SharedKeyCredential instead of connection string

    // Assuming there is some Azure::Core::IO::BodyStream containing the data to upload.
    // For example either a MemoryBodyStream or FileBodyStream.
    std::vector<uint8_t> data;
    Azure::Core::IO::MemoryBodyStream stream(data);

    Azure::Response<Models::UploadBlockBlobResult> response = block.Upload(stream);

    Models::UploadBlockBlobResult model = response.Value;

>     printf("Last modified date of uploaded blob: %s\n", model.LastModified.ToString().c_str());

Reviewer: Let's use `std::cout`, and then we also don't need the `c_str()`.

  }
  catch (const Azure::Core::RequestFailedException& e)
  {

>     printf("Status Code: %d, Reason Phrase: %s\n", (int)e.StatusCode, e.ReasonPhrase.c_str());

Reviewer: and also, we do mix and match `std::cout` and `printf`, and also we call `printf()` not as `std:printf()` - I don't like it.
Btw, do our headers do `using namespace std` or `include <stdio.h>` instead of `include <cstdio>` so that our code sample can call `printf()` withot an `std::`? That should not be happening.

Reviewer: Also, this statement uses a C-style cast. Should be `static_cast<int>(e.StatusCode)`.

    std::cout << e.what() << std::endl;
    return 1;
  }
  return 0;
}
```

#### Key Core concepts

> Understanding the key concepts from the `azure-core-cpp` library, which is leveraged by all client libraries will also be helpful in getting started, regardless of which Azure service you want to use.

Reviewer: Should we call the library `azure-core`, or even "Azure Core"? `azure-core-cpp` is vcpkg-specific package name under which we are available there.

> The main shared concepts of `Azure::Core` include:

Reviewer: And here we call it `Azure::Core`. Let's just use "Azure Core"? Not `azure-core`, not `azure-core-cpp`, not `Azure.Core`.

- Accessing HTTP response details for the returned model of any SDK client operation, via `Response<T>`.
- Exceptions for reporting errors from service requests in a consistent fashion via the base exception type `RequestFailedException`.
- Abstractions for Azure SDK credentials (`TokenCredential`).
- Handling streaming data and input/output (I/O) via `BodyStream` along with its derived types.

> - Polling long-running operations (LROs), via `Operation<T>`.

Reviewer 3: Add our paging support too?

Author: Yes! Good call out, we did finally add that.

- HTTP pipeline and HTTP policies such as retry and logging, which are configurable via service client specific options.
- Replaceable HTTP transport layer to send requests and receive responses over the network.

#### `Response <T>` Model Types

Many client library operations **return** the templated `Azure::Core::Response<T>` type from the API calls. This type let's you get the raw HTTP response from the service request call the Azure service APIs make, along with the result of the operation to get more API specific details. This is the templated `T` operation result which can be extracted from the response, using the `Value` field.

```C++
  std::string containerName = "testcontainer";

  BlobContainerClient containerClient

>       = BlobContainerClient::CreateFromConnectionString(GetConnectionString(), containerName);

Reviewer 3: Replace connection string part of this example

  // Azure service operations return a Response<T> templated type
  Azure::Response<Models::CreateBlobContainerResult> createResult
      = containerClient.CreateIfNotExists();

  // You can extract the T, from the returned Response<T>,
  // which is typically named with a Result suffix in the type name.

>   Models::CreateBlobContainerResult result = createResult.Value;

Reviewer 3: Here the variable is 'result' but earlier the variable is 'model' - can we make consistent.

  // Now you can look at API specific members on the result object that is returned.

>   printf("Did the container get created? %d\n", result.Created);

Reviewer: std::cout

Author: Do you know what's the best way to emit a bool Anton?
Can you please share a suggestion/code snippet.

Reviewer: `std::cout << "Did the container get created? " << std::boolalpha << result.Created << std::endl;`?
https://onlinegdb.com/SFgGxu_S-7

But I would not bother to put it, if C++ don't have it by default, plus applying `std::boolalpha` changes the behavior in all other statements in the program. Unless you set and restore ios flags, but all that diverts the attention. I'd say just do "`std::cout << "Did the container get created? " << result.Created << std::endl;`" - it is pretty good to illustrate the SDK without diverting attention.

>   printf("Did the container get created? %d\n", result.Created);

Reviewer 3: Perhaps CreateIfNotExists is a strange choice to demo. If creates it if it doesn't exist and doesn't if it does. So, unless it throws, the container does exist. So it is created which makes the printf text odd. I think a different function would be a better demo.

```

#### Long Running Operations

Some operations take a long time to complete and require polling for their status. Methods starting long-running operations return `Operation<T>` types.

You can intermittently poll whether the operation has finished by using the `Poll()` method on the returned `Operation<T>` and track progress of the operation using `Value()`. Alternatively, if you just want to wait until the operation completes, you can use `PollUntilDone()`.

```C++

>   SomeServiceClient client;

Reviewer 3: I think you need to say that how this is initialized is not shown here otherwise a copy/past of this code is guaranteed to fail.

Author: Agreed, do you think I should use a concrete/real example, like start copy from uri instead?

Reviewer 3: Yes

  auto operation = client.StartSomeLongRunningOperation();

  while (!operation.IsDone())
  {

>     Azure::Core::Http::RawResponse pollResponse = operation.Poll();

Reviewer 3: why not use auto here to make the code less pedantic

Reviewer 3: Also, I think it better to show PollUntilDone without the loop as this is the more common case and the code is much shorter.

Author: > why not use auto here to make the code less pedantic

I could do that, I was hoping to show more of the types/namespaces, but given `RawResponse` isn't something end users will care about right out the gate, I will use auto. Do you recommend I do that in a lot more places (auto everything) or just where it is not as relevant?

Reviewer 3: I think showing that the code is simple is important; so yes, I think auto everywhere is good.
